### PR TITLE
Support multiple instances of a class in GlobalizeObjectGraph.

### PR DESCRIPTION
**Pull request by silvasean**
This happens in practice with e.g. ResNet from torchvision (multiple
instances of the same BatchNorm class).

The key observation is that for this program, and the expected set of
programs, we can convert the program to the same globalized form with a
bit more static analysis and effort to suitably monomorphize the
program. Though what we are doing here is fairly annoying to implement,
it saves any nontrivial later pass from having to do similar analyses
(or worse). E.g. shape inference would need to be object-graph aware,
mutation/lifetime analyses would have to be aware, etc. Additionally, it
would make us front-load what it means to have a !torch.nn.Module type
on an ABI boundary, which we are just not ready to handle.

I'm really, really hoping that in practice we can get away with
this, otherwise it's going to be really rough designing a representation
(and implementing everything to back it) that is convenient to transform
and gracefully scales from full object graph (in the most dynamic case)
down to a fixed set of global slots like we have here (in the most
static case, which we presume a lot of practical programs fall into).

This also involved introducing a
`torch-prepare-for-globalize-object-graph` pass that does a minimal set of
lowerings to simplify the IR into a more orthogonal and analyzable form,
and a `torch-globalize-pipeline` helper.

Recommended review order:
- updated documentation in Passes.td
- new tests in `globalize-object-graph-multiple-instances*.mlir`
- implementation of GlobalizeObjectGraph.cpp
- PrepareForGlobalizeObjectGraph.cpp + prepare-for-globalize-object-graph.mlir
- misc stuff like torch-globalize-pipeline pipeline definition.

With this, we can import, globalize, and inline resnet18 from
torchvision:
https://gist.github.com/silvasean/821586afc19b67d9fb72030b2e0adeb8

### CODE REVIEW

**Comment by stellaraccident** [comment]
High level comment: The rationale here makes sense to me. It seems that in the eventual case, even if you support non-globalized instances, the nature of these programs is such that you are still going to want to do what you are doing here as an optimization. It is a nice situation that there are enough simple programs that legalize fully to this form to make early progress.

File: lib/Dialect/Torch/Transforms/GlobalizeObjectGraph.cpp [hunk 1]
```diff
@@ -16,258 +16,379 @@
...
       continue;
-    if (rootNnModule) {
-      op.emitError()
-          .append("found more than one root module (module that is not a "
-                  "child of any other module)")
-          .attachNote(rootNnModule.getLoc())
-          .append("see other root module here");
-      return failure();
+    Value instance = mapping.lookupOrNull(operand.value());
+    if (!instance) {
```
Comment (stellaraccident) [inline lib/Dialect/Torch/Transforms/GlobalizeObjectGraph.cpp:304]: Trivial braces.

**Review by stellaraccident — APPROVED** [review]
Impressive work. Thanks.

**Comment by silvasean** [comment]
> High level comment: The rationale here makes sense to me. It seems that in the eventual case, even if you support non-globalized instances, the nature of these programs is such that you are still going to want to do what you are doing here as an optimization. It is a nice situation that there are enough simple programs that legalize fully to this form to make early progress.

Exactly!
